### PR TITLE
Adds Longest Arithmetic Subsequence Implementation

### PR DESCRIPTION
**Pull request by tejaswi0910**
Description
This PR implements the Longest Arithmetic Subsequence algorithm. The algorithm identifies the length of the longest arithmetic subsequence from a given array of integers, where a subsequence is defined as arithmetic if the differences between consecutive elements are the same.

Key Changes:
Added _LongestArithmeticSubsequence.java_ under the _dynamicprogramming_ directory.
The implementation uses dynamic programming with a HashMap to store the differences between elements and keep track of subsequences efficiently.

- [ ] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [ ] This pull request is all my own work -- I have not plagiarized it.
- [ ] All filenames are in PascalCase.
- [ ] All functions and variable names follow Java naming conventions.
- [ ] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [ ] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

**Review by vil02 — CHANGES_REQUESTED** [review]
Please remove the `main` method and add proper junit tests instead. Prefer `ParameterizedTest`s.

**Comment by tejaswi0910** [comment]
> Please remove the `main` method and add proper junit tests instead. Prefer `ParameterizedTest`s.

sure @vil02 , removed the main function and added tests!

File: src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+     *
+     * A sequence seq is arithmetic if seq[i + 1] - seq[i] are all the same value
+     * (for 0 <= i < seq.length - 1).
+     *
+     * @param nums the input array of integers
+     * @return the length of the longest arithmetic subsequence
+     */
+    public static int getLongestArithmeticSubsequenceLength(int[] nums) {
+        // If the array is empty or has only one element, return its length.
+        if (nums == null || nums.length == 0) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java:21]: What do you think about throwing an exception if `nums` is `null`?

Please add a test, with the input being `null`.

File: src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            return 0;
+        }
+
+        if (nums.length == 1) {
+            return 1;
+        }
+
+        int n = nums.length;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java:29]:
```suggestion
        final int n = nums.length;
```
I am not sure if creating this variable is needed - there is not overhead of using `nums.length` directly.

File: src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        HashMap<Integer, Integer>[] dp = new HashMap[n];
+        int maxLength = 2;
+
+        // Initialize dp array
+        for (int i = 0; i < n; i++) {
+            dp[i] = new HashMap<>();
+            for (int j = 0; j < i; j++) {
+                int diff = nums[i] - nums[j];
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java:37]:
```suggestion
                final int diff = nums[i] - nums[j];
```

File: src/test/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequenceTest.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class LongestArithmeticSubsequenceTest {
+
+    @ParameterizedTest
+    @MethodSource("provideTestCases")
+    void testGetLongestArithmeticSubsequenceLength(int[] nums, int expected) {
+        assertEquals(expected, LongestArithmeticSubsequence.getLongestArithmeticSubsequenceLength(nums));
+    }
+
```
Comment (vil02) [inline src/test/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequenceTest.java:17]:
```suggestion

        @ParameterizedTest
    @MethodSource("provideTestCases")
    void testGetLongestArithmeticSubsequenceLengthReversedInput(int[] nums, int expected) {
        ArrayUtils.reverse(nums);
        assertEquals(expected, LongestArithmeticSubsequence.getLongestArithmeticSubsequenceLength(nums));
    }

```

Make sure to:
```java
import org.apache.commons.lang3.ArrayUtils;
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java (hunk 1 above)
Comment (tejaswi0910) [reply src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java:21]: agreed, throwing an exception for the null case will be better to handle the invalid input explicitly

File: src/test/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequenceTest.java (hunk 4 above)
Comment (vil02) [reply src/test/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequenceTest.java:21]: Where did the original test go?

File: src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+     * Returns the length of the longest arithmetic subsequence in the given array.
+     *
+     * A sequence seq is arithmetic if seq[i + 1] - seq[i] are all the same value
+     * (for 0 <= i < seq.length - 1).
+     *
+     * @param nums the input array of integers
+     * @return the length of the longest arithmetic subsequence
+     */
+    public static int getLongestArithmeticSubsequenceLength(int[] nums) {
+        // Throws an exception if nums is null.
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java:20]:
```suggestion
```

File: src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        if (nums == null) {
+            throw new IllegalArgumentException("Input array cannot be null");
+        }
+
+        // If the array is empty or has only one element, return its length.
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequence.java:25]:
```suggestion
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/test/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequenceTest.java (hunk 4 above)
Comment (tejaswi0910) [reply src/test/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequenceTest.java:31]: Oops! I might have omitted it while adding the suggested code, but I have added it back now!

File: src/test/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequenceTest.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    @Test
+    void testNullInput() {
+        // Verify that an IllegalArgumentException is thrown when nums is null
+        assertThrows(IllegalArgumentException.class, () -> LongestArithmeticSubsequence.getLongestArithmeticSubsequenceLength(null));
+    }
+
+    private static Stream<Arguments> provideTestCases() {
+        return Stream.of(Arguments.of(new int[] {3, 6, 9, 12, 15}, 5), Arguments.of(new int[] {1, 7, 10, 13, 14, 19}, 4), Arguments.of(new int[] {1, 2, 3, 4}, 4), Arguments.of(new int[] {}, 0), Arguments.of(new int[] {10}, 1), Arguments.of(new int[] {9, 4, 7, 2, 10}, 3));
```
Comment (vil02) [inline src/test/java/com/thealgorithms/dynamicprogramming/LongestArithmeticSubsequenceTest.java:33]: Please add a test case, when the arithmetic sequence is constant, e.g.:
```suggestion
        return Stream.of(Arguments.of(new int[] {3, 6, 9, 12, 15}, 5), Arguments.of(new int[] {1, 7, 10, 13, 14, 19}, 4), Arguments.of(new int[] {1, 2, 3, 4}, 4), Arguments.of(new int[] {}, 0), Arguments.of(new int[] {10}, 1), Arguments.of(new int[] {9, 4, 7, 2, 10}, 3),
            Arguments.of(new int[] {1, 2, 2, 2, 2, 5}, 4));
```

**Review by vil02 — CHANGES_REQUESTED** [review]

**Review by vil02 — APPROVED** [review]

**Comment by vil02** [comment]
Just a small tip: if you want to modify the last commit, you can use `git commit --amend` and then just push your changes with `--force`.


Happy hacktoberfest!

**Comment by tejaswi0910** [comment]
> Just a small tip: if you want to modify the last commit, you can use `git commit --amend` and then just push your changes with `--force`.
> 
> Happy hacktoberfest!

Thanks @vil02, didn't know this....will surely use this in the future!
